src/backend/distributed/executor/adaptive_executor.c

```diff
@@ -1749,9 +1759,51 @@ AcquireExecutorShardLocksForExecution(DistributedExecution *execution)
 			 * concurrently.
 			 */
 
-			LockRelationShardResources(task->relationShardList, ExclusiveLock);
+			/*
+			 * Due to PG commit 5ee190f8ec37c1bbfb3061e18304e155d600bc8e we copy the
+			 * second parameter in pre-13.
+			 */
+			requiresConsistentSnapshotRelationShardList =
+				list_concat(requiresConsistentSnapshotRelationShardList,
+#if (PG_VERSION_NUM >= PG_VERSION_12) && (PG_VERSION_NUM < PG_VERSION_13)
+
+							list_copy(task->relationShardList));
+#else
+							task->relationShardList);
+#endif
 		}
 	}
+
+	/*
+	 * Acquire the locks in a sorted way to avoid deadlocks due to lock
+	 * ordering across concurrent sessions.
+	 */
+	anchorShardIntervalList =
+		SortList(anchorShardIntervalList, CompareShardIntervalsById);
+
+	/*
+	 * If we are dealing with a partition we are also taking locks on parent table
+	 * to prevent deadlocks on concurrent operations on a partition and its parent.
+	 */
+	LockParentShardResourceIfPartition(anchorShardIntervalList, lockMode);
+
+	/* Acquire distribution execution locks on the affected shards */
+	SerializeNonCommutativeWrites(anchorShardIntervalList, lockMode);
+
+	/* Acquire additional locks for SELECT .. FOR UPDATE on reference tables */
+	AcquireExecutorShardLocksForRelationRowLockList(relationRowLockList);
+
+	/*
+	 * If the task has a subselect, then we may need to lock the shards from which
+	 * the query selects as well to prevent the subselects from seeing different
+	 * results on different replicas.
+	 *
+	 * ExclusiveLock conflicts with all lock types used by modifications
+	 * and therefore prevents other modifications from running
+	 * concurrently.
+	 */
+	LockRelationShardResources(requiresConsistentSnapshotRelationShardList,
```

Comment:
If wouldn't mind some `if (requiresConsistentSnapshotRelationShardList != NIL)` statements around these to make it clear that we're usually not taking all these different types of locks.

Reply:
done